src/eduid_webapp/email/views.py

```diff
@@ -0,0 +1,295 @@
+                             'for user {!r}'.format(email, user))
+    try:
+        mail = user.mail_addresses.find(email)
+    except IndexError:
+        current_app.logger.debug('Couldnt save email {!r} as primary for user'
+                                 ' {!r}, data out of sync'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'out_of_sync'}
+        }
+
+    if not mail.is_verified:
+        current_app.logger.debug('Couldnt save email {!r} as primary for user'
+                                ' {!r}, email unconfirmed'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'emails.unconfirmed_address_not_primary'}
+        }
+
+    user.mail_addresses.primary = mail.email
+    try:
+        save_dashboard_user(user, dbattr_name='dashboard_userdb')
+    except UserOutOfSync:
+        current_app.logger.debug('Couldnt save email {!r} as primary for user'
+                                 ' {!r}, data out of sync'.format(email, user))
+        return {
+            '_status': 'error',
+            'error': {'form': 'out_of_sync'}
+        }
+    current_app.logger.info('Email address {!r} made primary '
+                            'for user {!r}'.format(email, user))
+    emails = {'emails': user.mail_addresses.to_list_of_dicts()}
+    return EmailListPayload().dump(emails).data
+
+
+@email_views.route('/verify', methods=['POST'])
+@UnmarshalWith(VerificationCodeSchema)
+@MarshalWith(EmailResponseSchema)
+@require_dashboard_user
+def verify(user, code, email, csrf_token):
+    """
+    """
+    if session.get_csrf_token() != csrf_token:
+        abort(400)
+    current_app.logger.debug('Trying to save email address {!r} as verified '
+                             'for user {!r}'.format(email, user))
+    db = current_app.verifications_db
+    state = db.get_state_by_eppn_and_code(user.eppn, code)
+
+    timeout = current_app.config.get('EMAIL_VERIFICATION_TIMEOUT', 24)
+    if state.is_expired(timeout):
+        msg = "Verification code is expired: {!r}".format(state.verification)
+        current_app.logger.debug(msg)
+        return {
+            '_status': 'error',
+            'error': {'form': 'emails.code_expired'}
+        }
+
+    if email != state.verification.email:
+        msg = "Invalid verification code: {!r}".format(state.verification)
+        current_app.logger.debug(msg)
+        return {
+            '_status': 'error',
+            'error': {'form': 'emails.code_invalid'}
+        }
+
+    state.verification.is_verified = True
+    state.verification.verified_ts = datetime.datetime.now()
+
+    state.verification.verified_by = user.eppn
+
+    current_app.verifications_db.save(state)
+    msg = 'Saved verification code: {!r} '.format(state.verification)
+    current_app.logger.debug(msg)
+
+    other = current_app.dashboard_userdb.get_user_by_mail(email,
```

Comment:
I would be happier it this functionality was moved to a separate function so it can be thoroughly tested.